From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Petr Hodina <petr.hodina@protonmail.com>
Date: Fri, 12 Dec 2025 00:00:00 +0000
Subject: [PATCH] clk: qcom: rcg2: Fix mdss byte/pixel clk set_rate on SM8750

Updating M/N/D registers on a gated RCG and toggling the UPDATE bit can
fail with "rcg didn't update its configuration", leaving disp_cc_mdss_pclk0
misconfigured during msm_dsi_host_power_on().

When the RCG is off, write the configuration directly without hitting
UPDATE. When it appears on, use the force-enable sequence used by shared
RCGs so UPDATE succeeds against the DSI PHY parent.

Signed-off-by: Petr Hodina <petr.hodina@protonmail.com>
Signed-off-by: David Heidelberg <david@ixit.cz>
---
 drivers/clk/qcom/clk-rcg2.c | 67 +++++++++++++++++++++++++++++++++++--
 1 file changed, 65 insertions(+), 2 deletions(-)

diff --git a/drivers/clk/qcom/clk-rcg2.c b/drivers/clk/qcom/clk-rcg2.c
index e18cb8807d735..a18d2b9319670 100644
--- a/drivers/clk/qcom/clk-rcg2.c
+++ b/drivers/clk/qcom/clk-rcg2.c
@@ -77,6 +77,9 @@ static int clk_rcg2_is_enabled(struct clk_hw *hw)
 	return (cmd & CMD_ROOT_OFF) == 0;
 }
 
+static int clk_rcg2_set_force_enable(struct clk_hw *hw);
+static int clk_rcg2_clear_force_enable(struct clk_hw *hw);
+
 static u8 __clk_rcg2_get_parent(struct clk_hw *hw, u32 cfg)
 {
 	struct clk_rcg2 *rcg = to_clk_rcg2(hw);
@@ -1085,8 +1088,36 @@ static int clk_byte2_set_rate(struct clk_hw *hw, unsigned long rate,
 
 	for (i = 0; i < num_parents; i++) {
 		if (cfg == rcg->parent_map[i].cfg) {
+			int ret;
+
 			f.src = rcg->parent_map[i].src;
-			return clk_rcg2_configure(rcg, &f);
+
+			if (!clk_hw_is_enabled(hw)) {
+				ret = regmap_read(rcg->clkr.regmap, RCG_CFG_OFFSET(rcg), &cfg);
+				if (ret)
+					return ret;
+
+				ret = __clk_rcg2_configure(rcg, &f, &cfg);
+				if (ret)
+					return ret;
+
+				return regmap_write(rcg->clkr.regmap, RCG_CFG_OFFSET(rcg), cfg);
+			}
+
+			ret = clk_rcg2_set_force_enable(hw);
+			if (ret)
+				return ret;
+
+			ret = clk_rcg2_configure(rcg, &f);
+			if (ret)
+				goto byte2_clear_force;
+
+			ret = clk_rcg2_clear_force_enable(hw);
+			return ret;
+
+byte2_clear_force:
+			clk_rcg2_clear_force_enable(hw);
+			return ret;
 		}
 	}
 
@@ -1180,7 +1211,38 @@ static int clk_pixel_set_rate(struct clk_hw *hw, unsigned long rate,
 		f.m = frac->num;
 		f.n = frac->den;
 
-		return clk_rcg2_configure(rcg, &f);
+		if (!clk_hw_is_enabled(hw)) {
+			int ret;
+
+			ret = regmap_read(rcg->clkr.regmap, RCG_CFG_OFFSET(rcg), &cfg);
+			if (ret)
+				return ret;
+
+			ret = __clk_rcg2_configure(rcg, &f, &cfg);
+			if (ret)
+				return ret;
+
+			return regmap_write(rcg->clkr.regmap, RCG_CFG_OFFSET(rcg), cfg);
+		}
+
+		{
+			int ret;
+
+			ret = clk_rcg2_set_force_enable(hw);
+			if (ret)
+				return ret;
+
+			ret = clk_rcg2_configure(rcg, &f);
+			if (ret)
+				goto pixel_clear_force;
+
+			ret = clk_rcg2_clear_force_enable(hw);
+			return ret;
+
+pixel_clear_force:
+			clk_rcg2_clear_force_enable(hw);
+			return ret;
+		}
 	}
 	return -EINVAL;
 }
-- 
2.53.0
